Fix get balance

jnewbery opened the pull request:
#12953 inadvertently removed the functionality to call `getbalance "*" <int> <bool>` to get the wallet's balance with either minconfs or include_watchonly.

This restores that functionality (when `-deprecatedrpc=accounts`), and also makes it possible to call ``getbalance minconf=<int> include_watchonly=<bool>` when accounts are not being used.

jnewbery committed:
[wallet] GetBalance can take an isminefilter filter.

GetBalance() can now take an ismine filter, which is passed down to
GetAvailableCredit. This allows GetBalance to be used to get watch-only
balances.

jnewbery committed:
[RPC] [wallet] allow getbalance to use min_conf and watch_only without accounts.

jnewbery committed:
[wallet] deduplicate GetAvailableCredit logic

jnewbery committed:
[wallet] factor out GetAvailableWatchOnlyBalance()

jnewbery committed:
[wallet] Factor out GetWatchOnlyBalance()

jnewbery committed:
[wallet] GetBalance can take a min_depth argument.

jnewbery commented:
Feedback to address from @ryanofsky

Some thoughts looking at code:

- It seems like it would be good to raise an error if `IsDeprecatedRPCEnabled("accounts")` and the account argument is not null or "*". Otherwise the argument would be silently ignored.
- Documentation seems to say default minconf is 1, while implementation uses 0 or 1.
- Maybe could dedup caching code in `GetAvailableCredit` a little with something like

```
CAmount* cache = filter == ISMINE_SPENDABLE  ? nAvailableCreditCached : 
                 filter == ISMINE_WATCH_ONLY ? nAvailableWatchCreditCached :
                 nullptr;
```

jnewbery commented:
> I guess that would not work for min dept of 0 (unconfirmed).

Are you saying that we also need to check `pcoin->InMempool()` for coins with 0 depth?

jnewbery commented:
I believe that I've addressed all of @ryanofsky's feedback. Russ - can you rereview and confirm?

Need to address @jonasschnelli's comment. Jonas - can you please confirm my question above? Thanks!

ryanofsky commented on src/wallet/rpcwallet.cpp:
```diff
@@ -895,38 +905,35 @@ static UniValue getbalance(const JSONRPCRequest& request)
-                filter = filter | ISMINE_WATCH_ONLY;
-
-        return ValueFromAmount(pwallet->GetLegacyBalance(filter, nMinDepth, account));
+        if (!IsDeprecatedRPCEnabled("accounts") && *account != "") {
```
In commit "[RPC] [wallet] allow getbalance to use min_conf and watch_only"

I think the safest thing here would be to raise an error whenever accounts are deprecated and account is non-null, because previously `getbalance(null)`, `getbalance("*")`, and `getbalance("")` each did different things, and it doesn't seem good to silently ignore a requested behavior. Another slightly less safe but maybe acceptable option would be to change `""` to `"*"` on this line, since `getbalance("*")` and `getbalance(null)` at least were more similar than `getbalance("*")` and `getbalance("")`

Previously, and currently with the deprecated accounts feature enabled:

- `getbalance(null)` would return total balance
- `getbalance("*")` would return total balance with legacy accounting method
- `getbalance("account")` would return balance associated with a named account
- `getbalance("")` would return leftover balance not associated with any named accounts

ryanofsky replied:
https://github.com/bitcoin/bitcoin/pull/13566#discussion_r199161775

Also want to note that another approach we could take here would not be to deprecate the `account` argument at all, but instead just rename it to `label` and pass it down as an option to the GetBalance() and GetAvailableCredit methods where it could be used to filter txouts there by address label. This would be a simple code change, just adding `if (!label || *label == GetLabelName(txout.scriptPubKey))`.

ryanofsky reviewed (commented):
utACK c57f2009f65ffedc7531644243d5d7ddd786edde if getbalance("") is not treated like getbalance(null), (see comment).

It looks like we might have had tests for this previously which were deleted. It could be better in the future to avoid dropping tests for deprecated features until the features are actually removed, though it is sometimes nice to simplify tests.

ryanofsky reviewed (commented):
utACK 43937db46edab8c226ce190e185d7a3285b3b364, only change since last review is requiring "*" instead of "" to signify all accounts. Thanks for implementing this and all the changes.

jnewbery commented:
Thanks for the review @ryanofsky !

Empact commented:
This calls for a test, no?

jnewbery commented:
@Empact - test added. Review plz :slightly_smiling_face:

Empact commented on src/wallet/wallet.cpp:
```diff
@@ -2224,7 +2194,7 @@ CAmount CWallet::GetUnconfirmedWatchOnlyBalance() const
             const CWalletTx* pcoin = &entry.second;
             if (!pcoin->IsTrusted() && pcoin->GetDepthInMainChain() == 0 && pcoin->InMempool())
-                nTotal += pcoin->GetAvailableWatchOnlyCredit();
+                nTotal += pcoin->GetAvailableCredit(true, ISMINE_WATCH_ONLY);
```
nit: comment bool purpose?

jnewbery replied:
I don't think it's the project's coding style to always comment arguments (at least, I don't see that elsewhere in the codebase as the norm).

Empact commented:
Should `CWallet::GetUnconfirmedBalance` use the cache?

Empact commented on src/wallet/wallet.cpp:
```diff
@@ -1977,8 +1977,20 @@ CAmount CWalletTx::GetAvailableCredit(bool fUseCache) const
+    } else if (filter == ISMINE_WATCH_ONLY) {
+        cache = &nAvailableWatchCreditCached;
+        cache_used = &fAvailableWatchCreditCached;
+    }
```
nit: could move this into a `fUseCache` block, connected with 1991 below.

jnewbery replied:
No, `cache` and `cache_used` are updated below even if `fUsedCache` is not set.

Empact commented on src/wallet/rpcwallet.cpp:
```diff
@@ -895,38 +905,35 @@ static UniValue getbalance(const JSONRPCRequest& request)
+
+    isminefilter filter = ISMINE_SPENDABLE;
+    if (!request.params[2].isNull() && request.params[2].get_bool()) {
+        filter = filter | ISMINE_WATCH_ONLY;
```
nit: `|=`

Empact commented on src/wallet/rpcwallet.cpp:
```diff
@@ -4416,7 +4423,7 @@ static const CRPCCommand commands[] =
     { "wallet",             "encryptwallet",                    &encryptwallet,                 {"passphrase"} },
     { "wallet",             "getaddressinfo",                   &getaddressinfo,                {"address"} },
-    { "wallet",             "getbalance",                       &getbalance,                    {"account","minconf","include_watchonly"} },
+    { "wallet",             "getbalance",                       &getbalance,                    {"account|dummy","minconf","include_watchonly"} },
```
Maybe show `*` here instead of dummy?

jnewbery replied:
`dummy` is the name of the argument. `*` is the value that it needs to be set to.

Empact commented on src/wallet/rpcwallet.cpp:
```diff
@@ -895,38 +905,35 @@ static UniValue getbalance(const JSONRPCRequest& request)
+    }
+
+    if (!account_value.isNull()) {
 
```
nit: whitespace

Empact commented on src/wallet/rpcwallet.cpp:
```diff
@@ -853,8 +853,9 @@ static UniValue getbalance(const JSONRPCRequest& request)
     }
 
-    if (request.fHelp || (request.params.size() > 3 && IsDeprecatedRPCEnabled("accounts")) || (request.params.size() != 0 && !IsDeprecatedRPCEnabled("accounts")))
+    if (request.fHelp || (request.params.size() > 3 ))
```
nit: whitespace

Empact commented:
utACK 702ae1e

laanwj commented:
utACK 702ae1e21a09d8c31406839c4ea507c5fa276898

promag commented on src/wallet/wallet.h:
```diff
@@ -460,9 +460,8 @@ class CWalletTx : public CMerkleTx
     CAmount GetCredit(const isminefilter& filter) const;
     CAmount GetImmatureCredit(bool fUseCache=true) const;
-    CAmount GetAvailableCredit(bool fUseCache=true) const;
+    CAmount GetAvailableCredit(bool fUseCache=true, const isminefilter& filter=ISMINE_SPENDABLE) const;
```
nit, drop `&`?

promag commented on src/wallet/wallet.cpp:
```diff
@@ -2154,16 +2139,17 @@ void CWallet::ResendWalletTransactions(int64_t nBestBlockTime, CConnman* connman
             const CWalletTx* pcoin = &entry.second;
-            if (pcoin->IsTrusted())
-                nTotal += pcoin->GetAvailableCredit();
+            if (pcoin->IsTrusted() && pcoin->GetDepthInMainChain() >= min_depth) {
```
`IsTrusted()` already computes `GetDepthInMainChain()`. Suggestion:
```cpp
bool IsTrusted(int min_depth = 1)
```

promag replied:
Ignore my comment, I tried that refactor and it doesn't look good.

promag reviewed (commented):
utACK 702ae1e. ~~Consider last comment as a way to not degrade `GetBalance` performance.~~

jnewbery commented:
@Empact @promag - thanks for the reviews. I'll look at addressing nits in a follow-up.

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -1977,8 +1977,20 @@ CAmount CWalletTx::GetAvailableCredit(bool fUseCache) const
+    } else if (filter == ISMINE_WATCH_ONLY) {
+        cache = &nAvailableWatchCreditCached;
+        cache_used = &fAvailableWatchCreditCached;
+    }
```
Seems like a reasonable optimization, but I don't think it's required. Could be done in a separate PR.

jnewbery commented:
@Empact @promag - I've addressed the remaining nits in https://github.com/jnewbery/bitcoin/tree/13566_nits, but there isn't much in there. I'm not sure if there's much point in opening a new PR.

Empact commented:
Yeah I would leave it as-is at this point.

promag commented:
@jnewbery agree, not worth it.

jnewbery commented:
Agree. Thanks for the input. Sorry those changes didn't make it!
